dev/requirements.txt

```diff
@@ -1,5 +1,7 @@
-flake8==3.5.0
+pycodestyle==2.5.0
+flake8==3.7.9
 jira==1.0.3
 PyGithub==1.26.0
 Unidecode==0.04.19
-sphinx
+sphinx==2.3.1
+numpy==1.18.1
```

Comment:
I quickly googled and skimmed other projects about the requirements [here](https://data-flair.training/blogs/python-open-source-projects/) and [here](https://hackernoon.com/50-popular-python-open-source-projects-on-github-in-2018-c750f9bf56a0). I still think it's more usual to specify the range, rather than the specific version.

I am still not sure if it's right to pin the version yet. There's trade-off on pinning. I think Spark community is big enough to handle these issue from using the latest versions too. I would only pin the version when an issue difficult to fix is found.

Reply:
cc somewhat arbitrary committers who might be interested in here: @srowen, @holdenk, @dongjoon-hyun, @BryanCutler. WDYT about this?

Reply:
I understand the trade-off mentioned by @HyukjinKwon . Actually, I tried to pin once before and dropped my PR due to that. +1 for @HyukjinKwon 's advice, `I would only pin the version when an issue difficult to fix is found.`.

The only exception is `spark-rm/Dockerfile`. We need to use specific version and to manager explicitly there.

Reply:
@HyukjinKwon - When looking at project dependencies, there is an important distinction between projects that are used as libraries and projects that are used as stand-alone applications.

If your project is a library, then you know others are importing you alongside other dependencies too. To minimize the chance of transitive dependency conflicts, you want to be flexible in how you specify your dependencies.

When your project is a stand-alone application, you don't have to worry about such things. You can pin every dependency to a specific version to get the most predictable and reliable build and runtime behavior.

In our case, the Spark build environment is more akin to a stand-alone application than a library. We don't need to worry about downstream users struggling with dependency conflicts. We can get the most stable build behavior by pinning everything, and there is no downside as far as I can tell.

I'll use [Trio](https://github.com/python-trio/trio) as an example again to illustrate my point:
* Trio is a library that others will typically import alongside many other dependencies. So in [Trio's setup.py](https://github.com/python-trio/trio/blob/4d956a4ba51241ca5d22800fb1e0e5c36ba9bb47/setup.py#L81-L92) they are very flexible in how they specify their dependencies.
* Trio's test environment, on the other hand, is only used by Trio contributors. So Trio [locks down](https://github.com/python-trio/trio/blob/4d956a4ba51241ca5d22800fb1e0e5c36ba9bb47/test-requirements.txt) every test requirement using [pip-tools](https://github.com/jazzband/pip-tools).

Reply:
There's the trade-off here as well - it's the most stable but there might be multiple bugs existing. It will still mess developer's environment. Arguably there are not so many Python dedicated dev people in Spark community who fluently uses pyenv, virtualenv or conda. Most of them just use pip and local installation.

I quickly skimmed requirements.txt **for dev or docs** at [here](https://hackernoon.com/50-popular-python-open-source-projects-on-github-in-2018-c750f9bf56a0). I skimmed top 20 projects, and found 6 requirements.txt for dev. I found 6 instances:

- One was completely pinned.
- Two were partially pinned
- Three were not quite pinned.

I agree the dev envs more tend to specify the versions; however, I think it's still prevailing to don't pin.

Reply:
What about just using a virtualenv to isolate the effect on installed libraries - wouldn't that solve that part?

Reply:
It will. I am good with using the virtualenv in CI and docker image. I use Conda too in my local.

My concern is that pinning versions will either force the dev people to use virtualenv (or its friends) or mess their installed packages.

Reply:
And also ideally using latest versions will bring some bug fixes. I assume we'll start to update this file to use latest pinned versions if this PR is merged, which is virtually equivalent with not pinning.

Not pinning take the risk about unexpected build failures in Github Actions - not in Jenkins since the packages are installed once, which I don't think is very critical.

Reply:
Would it help then if we expanded `dev/README.md` to show how to setup a virtual environment? I'm willing to do that.

If we don't want to ask devs to use virtual environments at all, then perhaps we need to fork `dev/requirements.txt` and have a version that pins everything, for use in CI and releases, and a version that pins nothing, for use by devs who don't use virtual environments.

Another alternative is the compromise currently standing in this PR, with some versions specified as `major.minor.*`.

And yet another alternative (which I personally wouldn't favor, but I know it's common) is to Dockerize the whole development environment, but that's a lot of work.

Reply:
I would suggest a note in README to ensure people know how they can isolate this env if needed, and pinning minor version only? is that close enough for consensus?

Reply:
OK, I'll do that. I'll recommend that users create a virtual environment in `dev/README.md` and demonstrate how to do that. I'll also update the version specifiers to all be in the form of `major.minor.*`.

Reply:
Hmm, on second thought, perhaps the README should be left to a separate PR.  We already have advice on setting up a development environment in at least a couple of places, like [Useful Developer Tools](http://spark.apache.org/developer-tools.html) and [`docs/README.md`](https://github.com/apache/spark/blob/fae981e5f32e0ddb86591a616829423dfafb4ed0/docs/README.md#using-rbenv-and-pyenv).

Perhaps we should consolidate that advice over on the Useful Developer Tools page, since it fits it with the information already there.

Either that, or let's agree on some other approach to take. But I think we can defer any dev documentation changes to a follow-up PR.

Reply:
I also would prefer not to pin specific versions and agree with https://github.com/apache/spark/pull/27928#discussion_r395137725. It is good to have the community try with the latest to surface any issues, but we should be very clear what versions have been used in our CI, which could be from a virtualenv or just in a readme, so there is always an obvious fallback version.